src/opcua-connector/machineState.ts

```diff
@@ -0,0 +1,132 @@
+
+/**
+ * Listen to all values in the stream and apply the rules if one of them changed
+ * There is a little debounce to get all values who are in the same query interval
+ *
+ * @param streams stream of all opcua values
+ * @param rules rules from the settings how an when to emit events
+ * @param emitState callback when a state event needs to be emitted
+ * @param emitError callback when an error event needs to be emitted
+ */
+export const executeStateEmitter = (
+  streams: OpcuaStreams,
+  rules: Rules,
+  emitState: EmitStateHandler,
+  emitError: EmitErrorHandler,
+): void => {
+  combineLatest(streams)
+    .pipe(
+      debounceTime(10),
+      map<VariableStream, VariableStreamData>(toValues),
+      map((values) => Object.entries(rules).map(validateRule(values)).filter(notUndefined)),
+      map((events) => events.map(renderDescription)),
+      distinctUntilChanged(deepEqual),
+      tap(publishEvents(emitState, emitError)),
+    )
+    .subscribe()
+}
+
+/** convert the variants to the real values as any type */
+const toValues = (values: VariableStream): VariableStreamData =>
+  Object.entries(values).reduce(
+    (acc, [key, v]) => ({ ...acc, [key]: v.value }),
+    {},
+  ) as VariableStreamData
+
+/**
+ * Validate the given rule with the value stream
+ *
+ * It will create a function, executing the code giving in the settings
+ *
+ * WARNING: this code is not validating the settings for any vulnerability code.
+ *
+ */
+const validateRule =
+  (values: VariableStreamData) =>
+  ([name, rule]: [string, Rule]): MachineStateEventSource | undefined => {
+    const vars = Object.entries(values).reduce(
+      (acc, [name, value]) =>
+        acc + `var ${name} = ${typeof value == 'string' ? "'" + value + "'" : value}\n`,
+      '',
+    )
+    const expression = vars + 'return ' + rule.rule
+    try {
+      const res = new Function(expression)()
+      if (res === true) {
+        return {
+          name,
+          values,
+          rule,
+        } as MachineStateEventSource
+      }
+      return undefined
+    } catch (e) {
+      console.error(e, name, expression)
+      return undefined
+    }
+  }
+
+/**
+ * Call the given handler by the state of the event and the applied rule.
+ *
+ * the rule will set the generateError flag
+ */
+const publishEvents =
+  (emitState: EmitStateHandler, emitError: EmitErrorHandler) => (events: MachineStateEvents) =>
+    events.forEach((event) => {
+      if (event.generateError) {
+        emitError(uuid.v4(), event.state, event.stateDesc)
+      }
+      emitState(event.state, event.stateDesc)
+    })
+
+/** Typed filter to filter undefined entries in an array */
+const notUndefined = <T>(v: T | undefined): v is T => v !== undefined
+
+/** render event description from the rule and from the current values */
+const renderDescription = ({ values, rule, name }: MachineStateEventSource): MachineStateEvent => {
```

Comment:
Since there are quite some event formatting concerns in different locations, let's think about introducing a simple templating language like mustache or hbs in a future version. WDYT?